lib/common/indirection.h

```diff
@@ -18,14 +18,55 @@
 // Defines a smart pointer class template that's rather like std::unique_ptr<>
 // but further restricted, like a C++ reference, to be non-null when constructed
 // or assigned.  Users need not check whether these pointers are null.
-// Intended to be as invisible as possible.
+// Supports copy construction, too.
+// Intended to be as invisible as a reference, wherever possible.
 
 #include "../common/idioms.h"
+#include <type_traits>
 #include <utility>
 
 namespace Fortran::common {
 
-template<typename A> class Indirection {
+// The default case does not support (deep) copy construction and assignment.
+template<typename A, bool COPY = false> class Indirection {
+public:
+  using element_type = A;
```

Comment:
What is the purpose of element_type?

Reply:
Reflection.  I generally save template type parameters as nested typedefs in case it's necessary to extract them later in another piece of code via `typename decltype(x)::element_type`.  I use the name `element_type` since the STL uses that name for this same purpose in many class templates, including `unique_ptr`, of which Indirection is meant to be a better version.